WordPress/gutenberg
# Show warning on critical block removal

## What?


Fixes #40618.

(edited by @mcsf:)

Show a modal asking the user for confirmation when attempting to remove Post Content and Query blocks, but only in the site editor.

~I have questions regarding implementation, mainly: should this initial version be extensible? If so, perhaps it would be better to register block criticality, as well as the message to display, in the block metadata. That would mean that third parties would be able to start using it immediately.~ For now, keep everything private using Gutenberg's new lock/unlock APIs. Everything is implemented inside the `@wordpress/block-editor` package. The `@wordpress/edit-site` just needs to unlock `BlockRemovalWarningModal` and render it in its `Editor` component.

~Another thought I had was should the check for whether to show the prompt be part of the `removeBlocks` action instead?~ Yes, this works inside the `removeBlocks` call. We've introduced some dedicated logic, introduced a private action named `privateRemoveBlocks`, and this new logic now comes with an extra argument, `forceRemove`, that is enabled when the user confirms deletion via the modal.


## Testing Instructions





A) In the site editor, go to a template that has either Post Content or Query (both considered critical blocks) and try the following:

1. Select the critical block and press backspace.
2. Select the critical block, navigate to its toolbar options dropdown and click "Delete".
3. Try both actions above with more than one block selected, including the critical block.
4. In the list view, try deleting the block with backspace, and through the options dropdown.
5. Also in the list view, try the above with more than one block selected, including the critical block.
6. In all these situations, the modal should pop up asking for confirmation of critical block removal.
~7. For bonus points: check if I missed any block removal methods 😄~

B) Make sure that there is no change in behaviour in other editors. For instance:
1. Open a new post in the Post Editor.
2. Insert one of the targeted blocks, e.g. by pasting in `` 
3. Make sure you are able to delete it without seeing the modal.


### Testing Instructions for Keyboard


## Screenshots or screencast 

<img width="387" alt="51145-deleting-single" src="https://github.com/WordPress/gutenberg/assets/150562/42ff9669-cdcc-4342-a761-0b9d4947a3d6">
<img width="419" alt="51145-deleting-multiple" src="https://github.com/WordPress/gutenberg/assets/150562/beec51c0-2cf9-444e-9aa8-173b916b5076">
